// js-server/MongoDB.js
module.exports = {
	insertVideosToDB: function (videos) {
		insertVideos(videos);
	},

	findVideoFromDB: function(videoId, resfn){
		findVideo(videoId, resfn);
	},

	getUserFromDB: function(userId, resfn){
		getUser(userId,resfn);
	},

	addUserLikesToDB: function(userId, videoId){
		addUserLikes(userId, videoId);
	},

	removeUserLikedVideos: function(userId, videoId){
		removeUserLikedVideos(userId, videoId);
	}
};

var MongoClient = require('mongodb').MongoClient;

function connectToDB(){
	MongoClient.connect("mongodb://trendeo_admin:trendeo123@ds031982.mongolab.com:31982/trendeo_videos", function(err, db) {
		if(!err) {
			console.log("Connected to DB");
		}else{
			console.log("DB error: " + err);
			return;
		}
		var collection = db.collection('videos');
		var doc1 = {'hello':'doc1'};
		var doc2 = {'hello':'doc2'};
		var lotsOfDocs = [{'hello':'doc3'}, {'hello':'doc4'}];

		collection.insert(doc1);

		collection.insert(doc2, {w:1}, function(err, result) {});

		collection.insert(lotsOfDocs, {w:1}, function(err, result) {});

		collection.findOne({'hello':'doc1'}, function(err, item) {
			console.log(item);
		});

		// collection.findOne({ 'videoId' : video['videoId'] }, function(err, item){
		// 	if(item != undefined && item['videoId']!=undefined){
		// 		console.log("video already in DB: " + item);
		// 	}else{

		//console.log(err + " " + result);

	});
}

function insertVideos(videos){
	if(videos == undefined || videos.length == 0) return;
	MongoClient.connect("mongodb://trendeo_admin:trendeo123@ds031982.mongolab.com:31982/trendeo_videos", function(err, db) {
		if(!err) {
			//console.log("Connected to DB");
		}else{
			console.log("DB error: " + err);
			return;
		}

		var collection = db.collection('videos');

		
		collection.insert(videos, {w:1, keepGoing:true}, function(err, result) {
			if(err) console.log(err);
		});
	});
}

function findVideo(videoId, resfn){
	MongoClient.connect("mongodb://trendeo_admin:trendeo123@ds031982.mongolab.com:31982/trendeo_videos", function(err, db) {
		if(!err) {
			//console.log("Connected to DB");
		}else{
			console.log("DB error: " + err);
			return;
		}

		var collection = db.collection('videos');
		collection.findOne({ 'videoId' : videoId }, function(err, item){
			//console.log(item);
			resfn(item);
		});
	});
}


function getUser(userId, resfn){
	MongoClient.connect("mongodb://trendeo_admin:trendeo123@ds035702-a.mongolab.com:35702/trendeo_users", function(err, db) {
		if(!err) {
			//console.log("Connected to DB");
		}else{
			console.log("DB error: " + err);
			return;
		}

		var collection = db.collection('users');
		collection.findOne({ 'userId' : userId }, function(err, item){
			//console.log(item);
			if(item==null){
				var user = {'_id':userId, 'userId':userId, 'vidLiked':[], 'vidDissed':[]};
				resfn(user);
				collection.insert(user, {w:1}, function(err, result) {
				});
			}else{
				resfn(item);
			}
		});
	});
}


function addUserLikes(userId, videoId){
	MongoClient.connect("mongodb://trendeo_admin:trendeo123@ds035702-a.mongolab.com:35702/trendeo_users", function(err, db) {
		if(!err) {
			//console.log("Connected to DB");
		}else{
			console.log("DB error: " + err);
			return;
		}

		var collection = db.collection('users');
		collection.findOne({ 'userId' : userId }, function(err, item){
			console.log(item);
			if(item==null){
				var user = {'_id':userId, 'userId':userId, 'vidLiked':[], 'vidDissed':[]};
				collection.insert(user, {w:1}, function(err, result) {

				});
			}else{
				collection.update({'userId':userId}, {$addToSet:{vidLiked:videoId}}, {w:1}, function(err, result) {
					console.log(err + "\n" + result);
				});
			}
		});
	});

}


function removeUserLikedVideos(userId, videoId){

}
