// js-server/RedditVideos.js
module.exports = {
  getRedditJSON: function (subreddit) {
	return queueToJSON(subreddit);
  },

  getRedditVideoJSON: function (videoId, resfn){
	return getRedditVideoInfo(videoId, resfn);
  },

  addUserChannel: function(channel){
	addChannel(channel);
  }
};

var request = require("request");
var mongodb = require("./MongoDB.js");

var cachedlist = {};
var videolists = {};
var tempvideolists = {};
var sublast = {"all" : "null"};
var subpage = {};
var subtries = {};
var subreddits = ['all', 'all2', 'videos', 'gopro', 'youtubehaiku', 'bloopers', 'funnyvideos', 'fail', 'UnexpectedThugLife',
	'StandUpComedy', 'deepintoyoutube', 'ContagiousLaughter','LearnUselessTalents', 'musicvideos',
	'Music', 'listentothis', 'hiphopheads', 'trailers', 'sports' , 'nba', 'soccer', 'nfl','PublicFreakout', 'StreetFights', 'respectporn',
	'kidsafevideos'];

var banned = {'Music':true, 'hiphopheads':true,'leagueoflegends':true,'hearthstone':true,'smashbros':true,'tf2':true,
	'DotA2':true,'gamegrumps':true,'2007scape':true, 'wow':true, 'GrandTheftAutoV_PC':true,'Games':true,
	'Minecraft':true, 'KotakuInAction':true,'witcher':true,'GlobalOffensive':true,'MonsterHunter':true,'osugame':true,
	'PS4':true, 'pcgaming':true, 'CLG':true,'funhaus':true, 'pcmasterrace':true, 'totalwar':true,'starcraft':true,
	'Planetside':true,'pathofexile':true,'ffxiv':true,'SSBM':true,'KerbalSpaceProgram':true,'roosterteeth':true,'gaming':true,
	'mindcrack':true,'MortalKombat':true,'Yogscast':true,'nerdcubed':true,'Smite':true,'skyrim':true,'Metroid':true,
	'HaloOnline':true,'speedrun':true,'Guildwars2':true,'heroesofthestorm':true,'runescape':true,'Warthunder':true,
	'metalgearsolid':true,'ethoslab':true,'Shaboozey':true,'Cynicalbrit':true,'starcitizen':true, 'killingfloor':true,
	'bindingofisaac':true,'Diablo':true,'Xcom':true,'DayZ':true,'dayz':true,'fallout':true,'playrust':true,'SSBPM':true,
	'GrandTheftAutoV':true,'StreetFighter':true, 'Metal':true, 'The_Donald':true};
var MAX_PAGE = 5;
var MAX_TRIES = 5;
var requesting = false;

setInterval(function(){
  try{
	loadRedditVideos();
  }catch(err){
	console.log(err);
  }
}, 15*60*1000);

loadRedditVideos();

function loadRedditVideos() {
	for(var i = 0; i < subreddits.length; i++){
		if(videolists[subreddits[i]]==undefined)
			videolists[subreddits[i]] = [];
			tempvideolists[subreddits[i]] = [];
		sublast[subreddits[i]] = "null";
		subpage[subreddits[i]] = 0;
		subtries[subreddits[i]] = 0;		
		loadSubredditVideos(subreddits[i]);
	}
}

function loadSubredditVideos(subreddit) {
	var url = "https://www.reddit.com/r/" + subreddit + "/.json?limit=60&after=" + sublast[subreddit];
	if(subreddit=="all")
		url = "https://www.reddit.com/domain/youtube.com/.json?limit=60&after=" + sublast["all"];
	if(subreddit=="all2")
		url = "https://www.reddit.com/domain/youtu.be/.json?limit=60&after=" + sublast["all2"];

	request({
		url: url,
		json: true
	}, function(error, response, body) {
		try{
			if (!error && response.statusCode === 200) {
				reqRedditSuccess(subreddit, body);
			}else{
				reqRedditError(subreddit, response, error);
			}
		}catch(err){
			console.log(err + err.stack);
		}
	});
}


function reqRedditSuccess(subreddit, body, successFn, errorFn){
	var tempqueue = procRedditVideos(body, subreddit);
	subtries[subreddit] = 0;
	subpage[subreddit] += 1;
	if(subpage[subreddit] >= MAX_PAGE){
		videolists[subreddit] = removeDuplicates(tempvideolists[subreddit]);
		if(subreddit == "all"){
			videolists[subreddit] = removeBanned(videolists[subreddit]);
		}else if(subreddit == "all2"){
			videolists["all"] = videolists["all"].concat(removeBanned(videolists[subreddit]));
		}
		//mongodb.insertVideosToDB(arrayToKeyValue(tempvideolists[subreddit]));
		tempvideolists[subreddit] = [];
		sublast[subreddit] = "null";
		subpage[subreddit] = 0;

	}else{
		tempvideolists[subreddit] = tempvideolists[subreddit].concat(tempqueue);
		loadSubredditVideos(subreddit);
	}
}

function reqRedditError(subreddit, response, error) {
	var errstr = subreddit + ':\n   Error:' + error;
	if(response != undefined)
		errstr += "\n   Status:" + response.statusCode;
	console.log(errstr);

	if(subtries[subreddit]++ < MAX_TRIES)
		loadSubredditVideos(subreddit);
	else
		console.log("Max tries reached for " + subreddit);
}


function procRedditVideos(data, subreddit){
	tempqueue = [];
	tempdb = [];
	for(var i = 0; i < data.data.children.length; i++){
		var item = data.data.children[i];
		var vidurl = getURLFromItem(item);
		if(vidurl !== null && vidurl !== undefined){
			var vidtitle = item.data.title;
			var vidSubreddit = item.data.subreddit;
			var videoId;
			if(vidurl.indexOf('v=') > 0)
				videoId = vidurl.substring(vidurl.indexOf('v=')+2);
			else if(vidurl.indexOf('%2Fwatch%3Fv%3D') > 0){
				videoId = vidurl.substring(vidurl.indexOf('%2Fwatch%3Fv%3D')+14,vidurl.indexOf('%26feature%3Dshare'));
			}else if(vidurl.indexOf('youtu.be/') > 0) {
				videoId = vidurl.substring(vidurl.indexOf('youtu.be/')+9);
			}
			var video = [videoId, vidtitle, vidSubreddit];
			tempqueue.push(video);
		}
	}
	sublast[subreddit] = data.data.after;
	return tempqueue;
}

function getURLFromItem(item){
	var dom = item.data.domain;
	var vidurl;
	if(dom == "youtube.com" || "youtu.be"){
		vidurl = item.data.url;
		if(vidurl== null && vidurl==undefined && item.data.secure_media != null && item.data.secure_media != undefined){
			vidurl = item.data.secure_media.oembed.url;
		}else if(vidurl== null && vidurl==undefined && item.data.media != null && item.data.media != undefined){
			vidurl = item.data.media.oembed.url;
		}
	}
	return vidurl;
}

function arrayToKeyValue(videos){
	var temp = [];
	for(var i = 0; i < videos.length; i++){
		temp.push({'_id':videos[i][0], 'videoId':videos[i][0], 'vidTitle':videos[i][1], 'vidSubreddit':videos[i][2]});
	}
	return temp;
}

function keyValueToArray(video){
	return [video.videoId, video.videoTitle];
}


function addChannel(channel){
	if(videolists[channel]!=undefined) return;

	//add channel to the list of videos
}

//return JSON
function queueToJSON(subreddit){
	var queue = videolists[subreddit];
	return queue;
	
	// if(queue==undefined || queue.length==0) return "[]";
	// var str = "[";
	// for(var i = 0; i < queue.length; i++){
	// 	while(queue[i][1].indexOf("\"")!=-1) queue[i][1] = queue[i][1].replace('\"','');
	// 	str += "[\"" + queue[i][0]+"\",\""+queue[i][1]+ "\"],";
	// }
	// str = str.substring(0,str.length-1) + "]";
	// return str;
}

function getVideoInfo(videoId, resfn){
	mongodb.findVideoFromDB(videoId, function(video){
		if(video!=null)
			resfn(videoToJSON(keyValueToArray(video)));
		else
			getVideoInfo(videoId, resfn);
	});
}

function getRedditVideoInfo(videoId, resfn){
	var url = "https://www.reddit.com/api/info.json?url=https://www.youtube.com/watch?v=" + videoId + "&callback=?";
	request({
		url: url,
		json: true
	}, function (error, response, body) {
		var maxscore = -1;
		var video = null;
		//console.log(error + "\n" + body);
		try{
			for(var i = 0; i < body.data.children.length; i++){
				var item = body.data.children[i];
				var vidurl = getURLFromItem(item);
				if(vidurl !== null && vidurl !== undefined){
					if(item.data.score > maxscore){
						var vidtitle = item.data.title;
						var vidSubreddit = item.data.subreddit;
						var videoId = vidurl.substring(vidurl.indexOf('v=')+2);
						video = [videoId, vidtitle, vidSubreddit];
					}
				}
			}
		}catch(err){
			console.log(error + "\n" + err.stack);
		}
		if(video==null){
			getYTVideoInfo(videoId, resfn);
		}else{
			resfn(videoToJSON(video));
		}
	});
}

function getYTVideoInfo(videoId, resfn){
	resfn([videoId,'blah blah blah']);
}

function videoToJSON(video){
	return video;
	// var str = "[";
	// if(video==null){
	// 	str += "null,null";
	// }else if(video != undefined){
	// 	while(video[1].indexOf("\"")!=-1) video[1] = video[1].replace('\"','');
	// 	str += "\"" + video[0] + "\",\"" + video[1] + "\"";
	// }
	// str = str + "]";
	// return str;
}

//remove duplicates
function removeDuplicates(queue){
	var hashtable = {};
	var newqueue = [];
	for(var i = 0; i < queue.length; i++){
		if(hashtable[queue[i][0]] != "true"){
			newqueue.push(queue[i]);
			hashtable[queue[i][0]] = "true";
		}
	}
	return newqueue;
}

//remove banned subreddits
function removeBanned(queue){
	var newqueue = [];
	for(var i = 0; i < queue.length; i++){
		if(!isBanned(queue[i][2])){
			newqueue.push(queue[i]);
		}
	}
	return newqueue;
}

function isBanned(subreddit){
	if(banned[subreddit])
		return true;
	else
		return false;
}
