// js-server/RedditVideos.js
module.exports = {
  getRedditJSON: function (subreddit) {
	return queueToJSON(subreddit);
  },

  addUserChannel: function(channel){
	addChannel(channel);
  }
};

var request = require("request");

var videolists = {};
var tempvideolists = {};
var sublast = {"all" : "null"};
var subpage = {};
var subreddits = ['all', 'videos', 'gifs', 'youtubehaiku', 'deepintoyoutube', 'StandUpComedy', 'fail', 'gopro', 'ContagiousLaughter',
	'todayilearned', 'LearnUselessTalents','Music', 'listentothis', 'hiphopheads','sports' , 'nba', 'soccer', 'nfl',
	'UnexpectedThugLife', 'space', 'PublicFreakout', 'StreetFights', 'respectporn'];
var banned = {'Music':true, 'hiphopheads':true,'leagueoflegends':true,'hearthstone':true,'smashbros':true,'tf2':true,
	'DotA2':true,'gamegrumps':true,'2007scape':true, 'wow':true, 'GrandTheftAutoV_PC':true,'Games':true,
	'Minecraft':true, 'KotakuInAction':true,'witcher':true,'GlobalOffensive':true,'MonsterHunter':true,'osugame':true,
	'PS4':true, 'pcgaming':true, 'CLG':true,'funhaus':true, 'pcmasterrace':true, 'totalwar':true,'starcraft':true,
	'Planetside':true,'pathofexile':true,'ffxiv':true,'SSBM':true,'KerbalSpaceProgram':true,'roosterteeth':true,'gaming':true,
	'mindcrack':true,'MortalKombat':true,'Yogscast':true,'nerdcubed':true,'Smite':true,'skyrim':true,'Metroid':true,
	'HaloOnline':true,'speedrun':true,'Guildwars2':true,'heroesofthestorm':true,'runescape':true,'Warthunder':true,
	'metalgearsolid':true,'ethoslab':true,'Shaboozey':true,'Cynicalbrit':true,'starcitizen':true, 'killingfloor':true,
	'bindingofisaac':true,'Diablo':true,'Xcom':true,'DayZ':true,'dayz':true,'fallout':true,'playrust':true,'SSBPM':true,
	'GrandTheftAutoV':true,'StreetFighter':true};
var MAX_PAGE = 5;

setInterval(function(){
  try{
	loadRedditVideos();
  }catch(err){
	console.log(err);
  }
}, 30*1000);

loadRedditVideos();

function loadRedditVideos() {
	for(var i = 0; i < subreddits.length; i++){
		if(videolists[subreddits[i]]==undefined)
			videolists[subreddits[i]] = [];
			tempvideolists[subreddits[i]] = [];
		sublast[subreddits[i]] = "null";
		subpage[subreddits[i]] = 0;
		loadSubredditVideos(subreddits[i]);
	}
}

function loadSubredditVideos(subreddit){
	var url = "https://www.reddit.com/r/" + subreddit + "/.json?limit=60&after=" + sublast[subreddit];
	
	if(subreddit=="all")
		url = "https://www.reddit.com/domain/youtube.com/.json?limit=60&after=" + sublast["all"];

	getRedditVideos(url, subreddit);
}

function getRedditVideos(url, subreddit) {
	var jsonres;
	request({
		url: url,
		json: true
	}, function (error, response, body) {
		if (!error && response.statusCode === 200) {
			var tempqueue = procRedditVideos(body, subreddit);
			subpage[subreddit] += 1;
			if(subpage[subreddit] >= MAX_PAGE){
				videolists[subreddit] = removeDuplicates(tempvideolists[subreddit]);
				tempvideolists[subreddit] = [];
				sublast[subreddit] = "null";
				subpage[subreddit] = 0;
				console.log(subreddit + ' cache refreshed to ' + videolists[subreddit].length + ' videos..');
			}else{
				//add the tempqueue to the temp video list of subreddit
				tempvideolists[subreddit] = tempvideolists[subreddit].concat(tempqueue);
				loadSubredditVideos(subreddit);
			}
		}else{
			console.log('Error:' + error);
			//add max tries
			//loadSubredditVideos(subreddit);
		}
	});
}

function procRedditVideos(data, subreddit){
	tempqueue = [];
	for(var i = 0; i < data.data.children.length; i++){
		item = data.data.children[i];
		var dom = item.data.domain;
		if(dom == "youtube.com"){
			var vidurl = item.data.url;
			if(item.data.secure_media != null && item.data.secure_media != undefined){
				vidurl = item.data.secure_media.oembed.url;
			}else if(item.data.media != null && item.data.secure_media != undefined){
				vidurl = item.data.media.oembed.url;
			}
			if(vidurl !== null && vidurl !== undefined){
				var vidtitle = item.data.title;
				var subr = item.data.subreddit;
				tempqueue.push([vidurl.substring(vidurl.indexOf('v=')+2), vidtitle, subr]);
			}
		}
	}
	sublast[subreddit] = data.data.after;
	return tempqueue;
}

function addChannel(channel){
	if(videolists[channel]!=undefined) return;

	//add channel to the list of videos
}

//return JSON
function queueToJSON(subreddit){
	var queue = videolists[subreddit];
	if(queue==undefined || queue.length==0) return "[]";
	var str = "[";
	for(var i = 0; i < queue.length; i++){
		while(queue[i][1].indexOf("\"")!=-1) queue[i][1] = queue[i][1].replace('\"','');
		str += "[\"" + queue[i][0]+"\",\""+queue[i][1]+ "\"],";
	}
	str = str.substring(0,str.length-1) + "]";
	return str;
}

//remove duplicates
function removeDuplicates(queue){
	var hashtable = {};
	var newqueue = [];
	for(var i = 0; i < queue.length; i++){
		if(hashtable[queue[i][0]] != "true"){
			newqueue.push(queue[i]);
			hashtable[queue[i][0]] = "true";
		}
	}
	return newqueue;
}

//remove banned subreddits
function removeBanned(queue){
	var newqueue = [];
	for(var i = 0; i < queue.length; i++){
		if(!isBanned(queue[i][2])){
			newqueue.push(queue[i]);
		}
	}
	return newqueue;
}

function isBanned(subreddit){
	if(banned[subreddit])
		return true;
	else
		return false;
}
