// js-server/Users.js
module.exports = {
	getUserJSON: function(userId, resfn){
		getUser(userId, resfn);
	},

	updateUserLikedVideos: function(userId, updown, video){
		updateUserLikes(userId, updown, video);
	}
};

var mongodb = require("./MongoDB.js");
var request = require("request");

function getUser(userId, resfn){
	mongodb.getUserFromDB(userId, function(user){
		mongodb.getVideosFromDB(user.vidLiked, function(videos){
			resfn(keyValueToArray(videos));
		})
	});
}


function keyValueToArray(videos){
	var temp = [];
	for(var i = 0; i < videos.length; i++){
		temp.push([videos[i].videoId, videos[i].videoTitle]);
	}
	return temp;
}

function updateUserLikes(userId, updown, video){
	//var inputToken = 'CAAXUl7PaI4oBAFACSJVqXLF4ZAqAfZAoLD9KsAKFZBlIzPuTnR2WjdtVTaSJNIaRKxm20P7ZCNqHqZAQdVDcd5EobIGWf00f2HG5GIMCYIWDbcBWi4pQ3PBVD93ZCwHgSQFOaGQl0jeZBoIK19Q92GiDvy02gU8lzAeqqLbR5n9u7xgjOTimuHdmCtThh4NFGV42pyPHiOuyF2vbtQOLGT5iZCEZAR3bP6rcZD';
	// var verifyurl = '/debug_token?input_token=' + userId + '&access_token=1641122906121098|pgsOp7k0L7DfpHw3d6y64gDmJEY';
	// verifyurl = '/me?access_token=' + userId;

	var vurl = 'https://graph.facebook.com/me?access_token=' + userId;
	request({
		url: vurl,
		json: true
	}, function (error, response, body) {
		if(body && body.id){
			var user_Id = body.id;
			console.log('user authed ' + user_Id);
			if(updown=='up'){
				addUserLikedVideos(user_Id, video);
			}else{
				removeUserLikedVideos(user_Id, video);
			}
		}else{
			console.log('user error \n' + userId + "\n" + body);
		}
	});
	//https://graph.facebook.com/me?access_token=CAAXUl7PaI4oBAFACSJVqXLF4ZAqAfZAoLD9KsAKFZBlIzPuTnR2WjdtVTaSJNIaRKxm20P7ZCNqHqZAQdVDcd5EobIGWf00f2HG5GIMCYIWDbcBWi4pQ3PBVD93ZCwHgSQFOaGQl0jeZBoIK19Q92GiDvy02gU8lzAeqqLbR5n9u7xgjOTimuHdmCtThh4NFGV42pyPHiOuyF2vbtQOLGT5iZCEZAR3bP6rcZD
}

function addUserLikedVideos(userId, video){
	//console.log(userId + 'adding ' + video);
	mongodb.insertVideosToDB(arrayToKeyValue(video));
	mongodb.addUserLikesToDB(userId,video[0]);
}

function removeUserLikedVideos(userId, video){
	mongodb.removeUserLikedVideos(userId,video);
}

function arrayToKeyValue(video){
	return {'_id':video[0], 'videoId':video[0], 'videoTitle':video[1],'videoSubreddit':'all'};
}
