// js-server/Users.js
module.exports = {
	getUserJSON: function(accessToken, resfn){
		getUser(accessToken, resfn);
	},

	updateUserLikedVideos: function(accessToken, updown, video){
		updateUserLikes(accessToken, updown, video);
	}
};

var mongodb = require("./MongoDB.js");
var request = require("request");

function getUser(accessToken, resfn){

	var appId = '1641122906121098';
	var vurl = 'https://graph.facebook.com/debug_token?input_token=' + accessToken + '&access_token=' + appId + '|pgsOp7k0L7DfpHw3d6y64gDmJEY';
	request({
		url: vurl,
		json: true
	}, function (error, response, body) {
		//console.log(body);
		if(body && body.data && body.data.user_id && body.data.app_id==appId){
			var user_Id = body.data.user_id;
			console.log('user authed ' + user_Id);

			mongodb.getUserFromDB(user_Id, function(user){
				mongodb.getVideosFromDB(user.vidLiked, function(videos){
					resfn(keyValueToArray(videos));
				})
			});

		}else{
			console.log('user error \n' + accessToken + "\n" + body);
		}
	});
	
}


function keyValueToArray(videos){
	var temp = [];
	for(var i = 0; i < videos.length; i++){
		temp.push([videos[i].videoId, videos[i].videoTitle]);
	}
	return temp;
}

function updateUserLikes(accessToken, updown, video){
	//var inputToken = 'CAAXUl7PaI4oBAFACSJVqXLF4ZAqAfZAoLD9KsAKFZBlIzPuTnR2WjdtVTaSJNIaRKxm20P7ZCNqHqZAQdVDcd5EobIGWf00f2HG5GIMCYIWDbcBWi4pQ3PBVD93ZCwHgSQFOaGQl0jeZBoIK19Q92GiDvy02gU8lzAeqqLbR5n9u7xgjOTimuHdmCtThh4NFGV42pyPHiOuyF2vbtQOLGT5iZCEZAR3bP6rcZD';
	// var verifyurl = '/debug_token?input_token=' + accessToken + '&access_token=1641122906121098|pgsOp7k0L7DfpHw3d6y64gDmJEY';
	// verifyurl = '/me?access_token=' + accessToken;

	//var vurl = 'https://graph.facebook.com/me?access_token=' + accessToken;
	var appId = '1641122906121098';
	var vurl = 'https://graph.facebook.com/debug_token?input_token=' + accessToken + '&access_token=' + appId + '|pgsOp7k0L7DfpHw3d6y64gDmJEY';
	request({
		url: vurl,
		json: true
	}, function (error, response, body) {
		//console.log(body);
		if(body && body.data && body.data.user_id && body.data.app_id==appId){
			var user_Id = body.data.user_id;
			console.log('user authed ' + user_Id);
			if(updown=='up'){
				addUserLikedVideos(user_Id, video);
			}else{
				removeUserLikedVideos(user_Id, video);
			}
		}else{
			console.log('user error \n' + accessToken + "\n" + body);
		}
	});
	//https://graph.facebook.com/me?access_token=CAAXUl7PaI4oBAFACSJVqXLF4ZAqAfZAoLD9KsAKFZBlIzPuTnR2WjdtVTaSJNIaRKxm20P7ZCNqHqZAQdVDcd5EobIGWf00f2HG5GIMCYIWDbcBWi4pQ3PBVD93ZCwHgSQFOaGQl0jeZBoIK19Q92GiDvy02gU8lzAeqqLbR5n9u7xgjOTimuHdmCtThh4NFGV42pyPHiOuyF2vbtQOLGT5iZCEZAR3bP6rcZD
}

function addUserLikedVideos(userId, video){
	//console.log(userId + 'adding ' + video);
	mongodb.insertVideosToDB(arrayToKeyValue(video));
	mongodb.addUserLikesToDB(userId,video[0]);
}

function removeUserLikedVideos(userId, video){
	mongodb.removeUserLikedVideos(userId,video);
}

function arrayToKeyValue(video){
	return {'_id':video[0], 'videoId':video[0], 'videoTitle':video[1],'videoSubreddit':'all'};
}
